### ISSUE TITLE
What is the goal of this package?

### ISSUE THREAD

**Issue by tomasmcz** [ufal/neuralmonkey#8]
I'm not quite sure what are we trying to achieve here. What is the goal of this package? How does it differ from [tflearn](https://github.com/tflearn/tflearn) and similar frameworks? Are we writing something that is already done somewhere? If not, what is new here?

These questions should be clearly answered in the README, if we want anybody to use this.

**Comment by jindrahelcl** [comment]
I see this repo as ufal's codebase mainly for neural MT experiments and similar tasks. From my point of view, one of the main purposes of this repo is for us to be able to take a today's paper from arxiv, write a  new component that implements the things from the paper, replace existing components with the new one and run it.

**Comment by tomasmcz** [comment]
@jindrahelcl Why don't we just use (and contribute to) tflearn or other existing framework? I understand that at the beginning of the MMMT experiment there was no way to see what a monstrous codebase it will become. But what is the reason to put huge amounts of work into maintaining it now? I'm not saying there is not any, I just want to clarify it and advertise it.

**Comment by jindrahelcl** [comment]
I think the codebase would be equally large using tflearn. The code isn't just architecture of neural networks but it includes also (say what you want) nice way of managing experiments, among other things. I don't see how for example mixer, copynet or byte-pair encoding preprocessor would fit into tflearn.

Speaking of tflearn, their metric functions are classes as well. ;-*

**Comment by jlibovicky** [comment]
Here are the reasons why I believe it makes sense work on the package:

**for us:** It is an interesting experience and kind of challenge to create a well-maintainable software for conducting experiments which can be used by third parties. At least, I haven't done it before. Preparing a project for MT Marathon will be also very interesting.

**for out department:** It is useful to have an in-house solution for these kind of models, such that whoever want to anything in this direction can just knock the the door of the neighbouring office and get a face-to-face tutorial from someone who has written the package.

**for the mankind:** It provides a bit higher-level of abstraction that the one provided by tflearn. As @jindrahelcl said, the same package could have written over tflearn with approximately the same amount of code. The building blocks we use in out configuration are more coarse-grained than in tfLearn or in Lasagne and therefore hopefully more accessible for less experienced users..

If you agree, I would write the a paraphrase of the previous paragraph to the README file together with references to other tools, as "If you want to use TensorFlow, you may also consider using: ..." Is there something except tfLearn that we should reference?
